src/cyclefold/gadgets.rs

```diff
@@ -0,0 +1,586 @@
+use super::nova_circuit::FoldingData;
+
+use crate::{
+  gadgets::{
+    ecc::AllocatedPoint,
+    r1cs::{AllocatedR1CSInstance, AllocatedRelaxedR1CSInstance},
+  },
+  traits::{commitment::CommitmentTrait, Engine, ROCircuitTrait},
+};
+
+use bellpepper_core::{ConstraintSystem, SynthesisError};
+pub struct AllocatedFoldingData<E: Engine> {
+  pub U: AllocatedRelaxedR1CSInstance<E>,
+  pub u: AllocatedR1CSInstance<E>,
+  pub T: AllocatedPoint<E>,
+}
+
+impl<E: Engine> AllocatedFoldingData<E> {
+  pub fn alloc<CS: ConstraintSystem<<E as Engine>::Base>>(
+    mut cs: CS,
+    inst: Option<&FoldingData<E>>,
+    limb_width: usize,
+    n_limbs: usize,
+  ) -> Result<Self, SynthesisError> {
+    let U = AllocatedRelaxedR1CSInstance::alloc(
+      cs.namespace(|| "U"),
+      inst.map(|x| &x.U),
+      limb_width,
+      n_limbs,
+    )?;
+
+    let u = AllocatedR1CSInstance::alloc(cs.namespace(|| "u"), inst.map(|x| &x.u))?;
+
+    let T = AllocatedPoint::alloc(cs.namespace(|| "T"), inst.map(|x| x.T.to_coordinates()))?;
+    T.check_on_curve(cs.namespace(|| "T on curve"))?;
+
+    Ok(Self { U, u, T })
+  }
+
+  // TODO: Delete if not needed, or uncomment if this saves time
+  // pub fn absorb_in_ro<CS: ConstraintSystem<<E as Engine>::Base>>(
+  //   &self,
+  //   mut cs: CS,
+  //   ro: &mut E::ROCircuit,
+  // ) -> Result<(), SynthesisError> {
+  //   self.U.absorb_in_ro(cs.namespace(|| "absorb U"), ro)?;
+  //   self.u.absorb_in_ro(ro);
+  //   ro.absorb(&self.T.x);
+  //   ro.absorb(&self.T.y);
+  //   ro.absorb(&self.T.is_infinity);
+  //   Ok(())
+  // }
+}
+
+pub mod emulated {
+  use bellpepper::gadgets::Assignment;
+  use bellpepper_core::{
+    boolean::{AllocatedBit, Boolean},
+    num::AllocatedNum,
+  };
+
+  use super::*;
+
+  use crate::{
+    constants::{/*BN_N_LIMBS,*/ NUM_CHALLENGE_BITS, NUM_FE_IN_EMULATED_POINT, /*NUM_FE_FOR_RO*/},
+    gadgets::{
+      nonnative::{bignat::BigNat, util::f_to_nat},
+      utils::{
+        alloc_bignat_constant, alloc_one, alloc_zero, conditionally_select,
+        conditionally_select_bignat, le_bits_to_num, /*scalar_as_base,*/
+      },
+    },
+    traits::{CurveCycleEquipped, Dual, Group, ROConstantsCircuit},
+    RelaxedR1CSInstance,
+  };
+
+  use ff::Field;
+
+  #[derive(Clone)]
+  pub struct AllocatedPoint<E>
```

Comment:
I think we should avoid the `is_infinity` bit here. In order to properly constrain it, we need to evaluate the curve equation non-natively which will be very expensive. Instead, (and if the properties of the curve allow), we could define `(0,0)` as the identity and let the CycleFold circuit check that this holds.